### ISSUE TITLE
support xhr.upload.addEventListener and removeEventListener

### ISSUE THREAD

**Issue by beckend** [jameslnewell/xhr-mock#32]
does not seem to work yet

**Comment by jameslnewell** [comment]
Which version are you using? Are you using `xhr-mock@next`? Can you please provide a snippet of a minimal example that you'd expect to work.

**Comment by beckend** [comment]
```javascript
const xhr = new XMLHttpRequest();

const handleProgress = (evt) => {
    console.log(evt);
};

xhr.upload.addEventListener('progress', handleProgress);
xhr.upload.removeEventListener('progress', handleProgress);
```

using 2.x preview 7

**Comment by beckend** [comment]
also on the note with this snippet which does not work in xhr-mock, but does in browsers.

```javascript
xhr.addEventListener('loadend', handleLoadEnd);
```
if error occurs, this does fire(it should).

**Comment by jameslnewell** [comment]
Sorry, can you please include a full example including the mock setup and full xhr details.

It *should* work according to these [tests](https://github.com/jameslnewell/xhr-mock/blob/master/src/MockXMLHttpRequest.test.ts#L207)

**Comment by beckend** [comment]
the tests looks convincing, makes me wonder if i did something weird, will get back to this coming days.

**Comment by jameslnewell** [comment]
Hi @beckend. Have you managed to figure out whats going wrong?

**Comment by beckend** [comment]
@jameslnewell sorry it took a while.
https://github.com/beckend/xhr-mock/blob/xhr_debug/test/debug.test.ts
You can clone the branch and test, the `fetchXHR` is used in production and is working intended in a webrowser.

The tests should pass but do not, it's the problem i have described in this thread.

**Comment by jameslnewell** [comment]
Sorry I've taken so long to respond! I had some time on the train to investigate this today - thanks for your detailed repo!

`xhr-mock` wasn't emitting upload progress events because you were using an empty string for the body and it was evaluating to false. I've fixed the issue and released `xhr-mock@2.0.0-preview.10`. https://github.com/jameslnewell/xhr-mock/blob/master/src/MockXMLHttpRequest.ts#L221

re the `loadend` event, I think it depends on what type of error occurred. Is that right? https://xhr.spec.whatwg.org/#request-error-steps

Currently only one type of error handling is implemented and we don't have a way to tell `xhr-mock` which type of error handling to execute. What do you think of something like:

```js
import {RequestError, ResponseError, EndOfBodyError} from 'xhr-mock';

mock.get('/', () => Promise.reject(new ResponseError()));
```

**Comment by beckend** [comment]
looks fine to me, xhr was never easy to mock to begin with
